src/metabase/notification/send.clj

```diff
@@ -343,18 +354,16 @@
                             (catch InterruptedException _
                               (log/warn "Interrupted while waiting for notification executor to terminate"))))))
     (dotimes [_ pool-size]
-      (.submit ^ExecutorService executor
-               ^Callable (fn []
-                           (while (not (Thread/interrupted))
-                             (try
-                               (let [notification (take-notification! queue)]
-                                 (send-notification-sync! notification))
-                               (catch InterruptedException _
-                                 (log/info "Notification worker interrupted, shutting down")
-                                 (throw (InterruptedException.)))
-                               (catch Exception e
-                                 (log/error e "Error in notification worker")))))))
+      (start-worker!))
     (fn [notification]
+      ;; ensure there is always a worker to handle the notification
+      (while (< (.getActiveCount ^ExecutorService executor) pool-size)
+        (log/debugf "Not enough workers, starting a new one %d/%d"
+                    (.getActiveCount ^ExecutorService executor)
+                    pool-size)
+        (start-worker!)
+        ;; wait a bit for the worker to start
+        (Thread/sleep 10))
```

Comment:
Why do we need this wait? Does the active count of the executor not update synchronously?

How about instead doing `(dotimes [i (- pool-size (.getActiveCount executor))] ...) instead?

This is also proof against looping infinitely if something if the workers are instantly dying or something.

Reply:
nice, dotimes is better!

about the sleep, yea it's not updated synchronously.